doc/vdo-design.rst

```diff
@@ -267,127 +303,299 @@ slab depot. Each write request causes an entry to be made in the journal.
+new block mappings in that block are stable on storage, and journal block
+writes are all issued with the FUA bit set to ensure the recovery journal
+entries themselves are stable. The journal entry and the data write it
+represents must be stable on disk before the other metadata structures may
+be updated to reflect the operation. These entries allow the vdo device to
+reconstruct the logical to physical mappings after an unexpected
+interruption such as a loss of power.
 
 *Write Path*
 
-A write bio is first assigned a "data_vio," the request object which will
-operate on behalf of the bio. (A "vio," from Vdo I/O, is vdo's wrapper for
-bios; metadata operations use a vio, whereas submitted bios require the
-much larger data_vio.) There is a fixed pool of 2048 data_vios. This number
-was chosen both to bound the amount of work that is required to recover
-from a crash, and because measurements indicate that increasing it consumes
-more resources, but does not improve performance. These measurements have
-been, and should continue to be, revisited over time.
-
-Once a data_vio is assigned, the following steps are performed:
-
-1.  The bio's data is checked to see if it is all zeros, and copied if not.
-
-2.  A lock is obtained on the logical address of the bio. Because
-    deduplication involves sharing blocks, it is vital to prevent
-    simultaneous modifications of the same block.
-
-3.  The block map tree is traversed, loading any non-leaf pages which cover
-    the logical address and are not already in memory. If any of these
-    pages, or the leaf page which covers the logical address have not been
-    allocated, and the block is not all zeros, they are allocated at this
-    time.
+An application write bio follows these steps through vdo:
+
+1.  A data_vio is obtained from the data_vio pool and associated with the
+    application bio. If there are no data_vios available, the incoming bio
+    will block until a data_vio is available. This provides back pressure
+    to the application. The data_vio pool is protected by a spin lock.
+
+    The newly acquired data_vio is reset and the bio's data is copied into
+    the data_vio if it is a write and the data is not all zeroes. The data
+    must be copied because the application bio can be acknowledged before
+    the data_vio processing is complete, which means later processing steps
+    will no longer have access to the application bio. The application bio
+    may also be smaller than 4K, in which case the data_vio will have
+    already read the underlying block and the data is instead copied over
+    the relevant portion of the larger block.
+
+2.  A logical lock is obtained on the logical address of the bio. It is
+    vital to prevent simultaneous modifications of the same logical
+    address, because deduplication involves sharing blocks. This is not
+    managed like a typical lock, but is simply a hashtable where the key is
```

Comment:
"typical lock" -> "mutex or spin lock" maybe? "is simply a hashtable" -> "uses a hashtable" since the antecedent appears to be a lock, not the overall lock managing mechanism